tapdb/universe.go

```diff
@@ -112,6 +113,243 @@ type BaseUniverseStore interface {
 		arg DeleteMultiverseLeaf) error
 }
 
+// specifierToIdentifier converts an asset.Specifier into a universe.Identifier
+// for a specific proof type.
+//
+// NOTE: This makes an assumption that only specifiers with a group key are
+// valid.
+func specifierToIdentifier(spec asset.Specifier,
+	proofType universe.ProofType) (universe.Identifier, error) {
+
+	var id universe.Identifier
+
+	// The specifier must have a group key to be able to be used within the
+	// ignore or burn tree context.
+	if !spec.HasGroupPubKey() {
+		return id, fmt.Errorf("group key must be set for proof type %v",
+			proofType)
+	}
+
+	id.GroupKey = spec.UnwrapGroupKeyToPtr()
+	id.ProofType = proofType
+
+	return id, nil
+}
+
+// getUniverseTreeSum retrieves the sum of a universe tree specified by its
+// identifier.
+func getUniverseTreeSum(ctx context.Context, db BatchedUniverseTree,
+	id universe.Identifier) universe.SumQueryResp {
+
+	namespace := id.String()
+	var sumOpt lfn.Option[uint64]
+
+	readTx := NewBaseUniverseReadTx()
+	txErr := db.ExecTx(ctx, &readTx, func(dbtx BaseUniverseStore) error {
+		tree := mssmt.NewCompactedTree(
+			newTreeStoreWrapperTx(dbtx, namespace),
+		)
+
+		// Get the root of the tree to retrieve the sum.
+		root, err := tree.Root(ctx)
+		if err != nil {
+			return err
+		}
+
+		// If root is empty, return empty sum.
+		if root.NodeHash() == mssmt.EmptyTreeRootHash {
+			return nil
+		}
+
+		// Return the sum from the root.
+		sumOpt = lfn.Some(root.NodeSum())
+		return nil
+	})
+	if txErr != nil {
+		return lfn.Err[lfn.Option[uint64]](txErr)
+	}
+
+	// If sumOpt was never set (empty tree), return None explicitly.
+	if !sumOpt.IsSome() {
+		return lfn.Ok(lfn.None[uint64]())
+	}
+
+	return lfn.Ok(sumOpt)
+}
+
+// uniKey is a type alias for a 32-byte array used as a key in the universe
+// tree.
+type uniKey = [32]byte
+
+// queryUniverseLeavesAndProofs executes a query against universe leaves,
+// fetches their inclusion proofs, and builds authenticated results.
+//
+// The LeafType is the concrete type of the leaf, AuthType is the type of the
+// wrapper of the LeafType that includes MS-SMT merkle proof info, and finally
+// the QueryType is the type that is used to query the leaves.
+func queryUniverseLeavesAndProofs[LeafType any, AuthType any, QueryType any](
+	ctx context.Context, db BatchedUniverseTree, assetSpec asset.Specifier,
+	id universe.Identifier,
+
+	// queryFunc retrieves the relevant UniverseLeaf records based on the
+	// specific query parameters.
+	queryFunc func(context.Context, BaseUniverseStore, asset.Specifier,
```

Comment:
Should we extract these function definitions into their own types to a) shorten this and b) have their descriptive comments moved to the type. Since these inline function parameter comments are really unusual for any of our repos.

Reply:
Sure. Earlier I went back and forth between making this a struct w/ an interface vs the closure approach. Ended up with the closure approach.